haskell/protobuf.bzl

```diff
@@ -0,0 +1,246 @@
+    # proto_library rule is in "foo/BUILD" but the .proto file is
+    # "foo/bar/baz.proto".
+
+    if not src.short_path.startswith(src_prefix):
+      fail("Mismatch between rule context " + str(ctx.label.package)
+           + " and source file " + src.short_path)
+    if src.basename[-6:] != ".proto":
+      fail("bad extension for proto file " + src)
+
+    relative_path = src.short_path
+
+    args.add([relative_path])
+
+    hs_files.append(ctx.actions.declare_file(
+      _proto_lens_output_file(relative_path),
+    ))
+
+  args.add([
+    "--haskell_out=no-reexports:" + paths.join(
+      hs_files[0].root.path,
+      src_prefix,
+    ),
+  ])
+
+  ctx.actions.run(
+    inputs = depset([
+      protobuf_tools(ctx).protoc,
+      protobuf_tools(ctx).protoc_gen_haskell,
+    ] + inputs),
+    outputs = hs_files,
+    executable = protobuf_tools(ctx).protoc,
+    arguments = [args],
+  )
+
+  hs_files_depset = depset(hs_files)
+
+  # Collect the generated source files from proto rule's dependencies.
+  for dep in ctx.rule.attr.deps:
+    hs_files_depset = depset(transitive = [
+      hs_files_depset,
+      dep[HaskellProtobufInfo].source_files,
+    ])
+
+  return [HaskellProtobufInfo(
+    source_files = hs_files_depset,
+  )]
+
+haskell_proto_aspect = aspect(
+  _haskell_proto_aspect_impl,
+  attr_aspects = ['deps'],
+  toolchains = ["@io_tweag_rules_haskell//protobuf:toolchain"],
+)
+
+def _protobuf_toolchain_impl(ctx):
+  return [
+    platform_common.ToolchainInfo(
+      name = ctx.label.name,
+      tools = struct(
+        protoc = ctx.executable.protoc,
+        protoc_gen_haskell = ctx.executable.protoc_gen_haskell,
+      ),
+      haskell_deps = ctx.attr.haskell_deps,
+      haskell_prebuilt_deps = ctx.attr.haskell_prebuilt_deps,
+    )
+  ]
+
+_protobuf_toolchain = rule(
+  _protobuf_toolchain_impl,
+  host_fragments = ["cpp"],
+  attrs = {
+    "protoc": attr.label(
+      executable = True,
+      cfg = "host",
+      allow_single_file = True,
+      mandatory = True,
+      doc = "protoc compiler",
+    ),
+    "protoc_gen_haskell": attr.label(
+      executable = True,
+      cfg = "host",
+      allow_single_file = True,
+      mandatory = True,
+      doc = "proto-lens-protoc plugin for protoc",
+    ),
+    "haskell_deps": attr.label_list(
+      doc = "List of other Haskell libraries to be linked to protobuf libraries.",
+    ),
+    "haskell_prebuilt_deps": attr.string_list(
+      doc = "Non-Bazel supplied Cabal dependencies for protobuf libraries.",
+    ),
+  },
+)
+
+def haskell_proto_toolchain(
+    name,
+    protoc_gen_haskell,
+    haskell_deps=[],
+    haskell_prebuilt_deps=[],
+    protoc=Label("@com_google_protobuf//:protoc"),
+    **kwargs):
+  """Declare a Haskell protobuf toolchain.
+
+  You need at least one of these declared somewhere in your `BUILD` files
+  for the `haskell_proto_library` rules to work. Once declared, you then
+  need to *register* the toolchain using `register_toolchain` in your
```

Comment:
Hm, it looks like most `X_proto_library` rules hard-code `@com_google_protobuf//:protoc`:
https://github.com/bazelbuild/bazel/blob/56eb80b11f475a2e5b6e8803aeb3cf2294e329b4/src/main/java/com/google/devtools/build/lib/rules/proto/ProtoConfiguration.java#L74

Which is also documented here:
https://blog.bazel.build/2017/02/27/protocol-buffers.html

So given that `rules_go` is also hard-coding `protoc` as an attribute default value, I guess it would be simpler for us to just do that too.

(However, I do think the `proto-lens-protoc` executable *should* be provided via a toolchain, since we don't have a consistent way of providing third-party packages.)

Reply:
But that link says `defaultValue`. We also have the same default value for `protoc` argument of the toolchain.

Reply:
@judah I think I understand the logic. You're saying that so long as we don't have any standard way to name haskell dependencies, it doesn't make sense to hard-code any label for `proto-lens-protoc`? I can buy that.

But how about this: add `@proto-lens-protoc` to `haskell_repositories()`, effectively vendoring in the code. Many `X_proto_library` rules rely on some amount of language-specific code. Including `rules_go`, which likewise vendors in.

Eventually, we'll want to strip the Nixpkgs dependency for building `@proto-lens-protoc`, to keep the build entirely self-contained (once we know how to do that). Even if the user still wants to rely on Nixpkgs for her own dependencies, it would be fine to build `proto-lens-protoc` in some completely different way. After all it's just a binary. And an internal implementation detail of `rules_haskell`.